### PR TITLE
Make NodeSamples editable

### LINKED ISSUE #20595: Hitsound sample changes aren't being applied to slider heads in the editor using osu! classic (2013) skin
### Type

Game behaviour

### Bug description

On the editor, when trying to switch the sample from normal to soft, it is only applied to hitcircles and slider ends - the slider head is left unchanged.

### Screenshots or videos

_No response_

### Version

2022.923.1-lazer

### Logs

[updater.log](https://github.com/ppy/osu/files/9727317/updater.log)
[database.log](https://github.com/ppy/osu/files/9727318/database.log)
[performance.log](https://github.com/ppy/osu/files/9727319/performance.log)
[runtime.log](https://github.com/ppy/osu/files/9727320/runtime.log)

### LINKED ISSUE #22187: Volume in editor doesn't affect slider heads or ticks
### Type

Game behaviour

### Bug description

Changing the volume of notes in the editor doesn't affect the slider head's or any sliderticks. It does correctly affect the slider's tail, the sliderslide, and sliderwhistles

Also, this might be a separate issue, and it's very strange, but the editor cannot add/remove hitwhistles (sliderwhistles on the sliderbody work correctly), finishes, or claps on sliders that already existed when the editor loads a beatmap, and it removes those samples, but only from slider heads, when the editor closes (it doesn't play the sample on slider heads when you test the beatmap, but when you go back to the editor the samples continue to play on the sliderheads until you exit the editor)

These issues aren't just aesthetic in the editor, any hitsounding done in the lazer editor will be messed up in this manner, even when the map is played normally (in the lazer client) or when exported to stable (in fact, all the hitsounds done in the editor disappear when exported except for those on slider bodies or tails)

### Screenshots or videos

https://user-images.githubusercontent.com/96793708/212401730-b8ece861-54e5-4514-a6fa-d8734cffa3c5.mp4

### Version

2022.1228.0

### Logs

[network.log](https://github.com/ppy/osu/files/10414779/network.log)
[performance.log](https://github.com/ppy/osu/files/10414780/performance.log)
[input.log](https://github.com/ppy/osu/files/10414783/input.log)
[database.log](https://github.com/ppy/osu/files/10414785/database.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/10414786/legacy-ipc.log)
[runtime.log](https://github.com/ppy/osu/files/10414790/runtime.log)
[updater.log](https://github.com/ppy/osu/files/10414791/updater.log)

### PR DESCRIPTION
**Pull request by OliBomby**
This change allows editing and modding of the `IHasRepeats.NodeSamples`. Also `SamplePointPiece` has been upgraded to include a 'addition bank', achieving parity with stable.

Still can't edit the custom sample index for legacy samples though. Not sure if we want to include that in the editor.

This checks two boxes in https://github.com/ppy/osu/issues/12020
Closes #22187
Closes #20595

osu Game Rulesets Osu Tests_y2uv5bNj0B

### CODE REVIEW

**Comment by Speykious** [comment]
Given this new sample bank menu:
image

I guess this now needs to be a dropdown.
image

**Comment by Speykious** [comment]
The right click menu on a hitobject would need to show the two sample banks as well, right?
image

**Comment by peppy** [comment]
@OliBomby hoping to get to this soon, but just to confirm forward direction (I think we may have discussed this so forgive me):

This is adding parity back, but the end goal will be to split the "addition bank" to be applied on a per-sample level, correct? This will have implications on the UI when we get to that point (basically the stuff in this PR will be replaced), so I'm curious if you already explored that direction and confirmed it's going to be a much larger scoped change before we go ahead with this smaller piece.

**Comment by peppy** [comment]
A few initial things:

Please check https://github.com/ppy/osu/pull/23443/commits/114f12a79056cc20951d06aec1ade073899d684a. I've moved the local logic of deciding which bank to use on creating new additions to the common method, since it felt awkward doing the same thing in multiple places when this method was intended to cover those cases.

---

Maybe a limitation of the implementation, but attempting to change the addition bank when there are no additions present will reset it back to the "normal" bank. Either this needs to be fixed, or the text box should be disabled when there are not yet any additions. I'd probably go with the latter as it feels like a better UX.


https://github.com/ppy/osu/assets/191335/368d9525-7302-479d-80fa-27d4fa29e0cb




---

There's no test coverage of the new behaviour. It'd be nice to have at least a bit.

**Review by peppy — CHANGES_REQUESTED** [review]
Initial pass, see attached feedback.

**Comment by OliBomby** [comment]
> @OliBomby hoping to get to this soon, but just to confirm forward direction (I think we may have discussed this so forgive me):
> 
> This is adding parity back, but the end goal will be to split the "addition bank" to be applied on a per-sample level, correct? This will have implications on the UI when we get to that point (basically the stuff in this PR will be replaced), so I'm curious if you already explored that direction and confirmed it's going to be a much larger scoped change before we go ahead with this smaller piece.

Yes, the intention is to eventually replace this UI so you can apply banks on a per-sample level. The reason I didn't go that direction yet is because the file encoding hasn't quite caught up yet, so saving banks on a per-sample level isn't possible with the legacy encoder.

The purpose of this PR is to bring parity with stable so people can edit hitsounds just like on stable (minus the custom samples).

**Comment by peppy** [comment]
Cool, we're on the same page then.

Make sure to review my follow-up comment with some actionable feedback.

**Comment by OliBomby** [comment]
> I guess this now needs to be a dropdown.

@Speykious I guess making it a dropdown is alright, since the purpose right now is only to bring parity with stable. Though eventually it will need to be a textbox again.

> The right click menu on a hitobject would need to show the two sample banks as well, right?

I don't think this bank context menu is necessary in the first place. You can just use the hitsound sample point thingy below the object on the timeline.

> Please check https://github.com/ppy/osu/commit/114f12a79056cc20951d06aec1ade073899d684a. I've moved the local logic of deciding which bank to use on creating new additions to the common method, since it felt awkward doing the same thing in multiple places when this method was intended to cover those cases.

@peppy Thank you for reviewing my PR! I agree with the changes, this is much cleaner.

> Maybe a limitation of the implementation, but attempting to change the addition bank when there are no additions present will reset it back to the "normal" bank. Either this needs to be fixed, or the text box should be disabled when there are not yet any additions. I'd probably go with the latter as it feels like a better UX.

Yeah it's impossible to add an additions bank when there are no additions to add the bank to. I agree with disabling the option in such case.

> There's no test coverage of the new behaviour. It'd be nice to have at least a bit.

I'll add some simple test cases to make sure that using these controls actually updates the hitsounds.

@peppy Should I invert the colours of the sample point pieces on sliders? so the pieces on node samples are pink and the piece for the main hitobject `Samples` (controlling the sliderbody hitsound) is purple. I feel like that would be more intuitive since for sliders and circles the pink sample point pieces would then have the same practical effect.

**Comment by peppy** [comment]
> Should I invert the colours of the sample point pieces on sliders? 

Can you take a screenshot showing what you mean?

**Comment by OliBomby** [comment]
> Can you take a screenshot showing what you mean?

osu Game Rulesets Osu Tests_mgrcCdqpPP
^ Current implementation

osu Game Rulesets Osu Tests_nmPXprgFrg
^ Suggested change

I'm aware that there are always two sample point pieces overlapping at the sliderhead. I'm not really sure what to do about that. You could make them narrower and put them next to eachother like in my follow-up PR, or merge them into one piece somehow, or just remove the piece for the sliderbody bank entirely.

**Comment by OliBomby** [comment]
I just realized https://github.com/ppy/osu/commit/114f12a79056cc20951d06aec1ade073899d684a broke some special logic in `SamplePointPiece`. It's actually supposed to prioritize inheriting the bank from the specific nodesample instead of the samples of the hitobject. If you only have `h.CreateHitSampleInfo(sampleName)` it will fetch the bank from the sliderbody hitsound when adding an addition to a slider repeat which is wrong.

File: osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs [hunk 1]
```diff
@@ -144,54 +201,222 @@ protected override void LoadComplete()
                 ScheduleAfterChildren(() => GetContainingInputManager().ChangeFocus(volume));
             }
 
-            private static string? getCommonBank(IList<HitSampleInfo>[] relevantSamples) => relevantSamples.Select(GetBankValue).Distinct().Count() == 1 ? GetBankValue(relevantSamples.First()) : null;
-            private static int? getCommonVolume(IList<HitSampleInfo>[] relevantSamples) => relevantSamples.Select(GetVolumeValue).Distinct().Count() == 1 ? GetVolumeValue(relevantSamples.First()) : null;
+            private string? getCommonBank() => relevantSamples.Select(GetBankValue).Distinct().Count() == 1 ? GetBankValue(relevantSamples.First()) : null;
+            private string? getCommonAdditionBank() => relevantSamples.Select(GetAdditionBankValue).Distinct().Count() == 1 ? GetAdditionBankValue(relevantSamples.First()) : null;
+            private int? getCommonVolume() => relevantSamples.Select(GetVolumeValue).Distinct().Count() == 1 ? GetVolumeValue(relevantSamples.First()) : null;
+
+            private void update(Action<HitObject, IList<HitSampleInfo>> updateAction)
```
Comment (peppy) [inline osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:208]: This whole flow is very hard to follow. I'm hoping we can find a way to simplify this, at least not passing an `Action` inside as it's done here.

File: osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs [hunk 2]
```diff
@@ -144,54 +201,222 @@ protected override void LoadComplete()
...
+                if (string.IsNullOrEmpty(sampleName))
+                    return;
+
+                update((h, samples) =>
+                {
+                    // Make sure there isn't already an existing sample
+                    if (samples.Any(s => s.Name == sampleName))
+                        return;
+
+                    // First try inheriting the sample info from the node samples instead of the samples of the hitobject
```
Comment (peppy) [inline osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:355]: I had a lot of trouble understanding how this changed anything until I realised that the `samples` list here is actually augmented from the `HitObject`'s samples.

File: osu.Game.Tests/Visual/Editing/TestSceneTimelineHitObjectBlueprint.cs [hunk 3]
```diff
@@ -111,5 +117,148 @@ public void TestDisallowRepeatsOnZeroDurationObjects()
...
+            {
+                var bankTextBox = popover.ChildrenOfType<IndeterminateSliderWithTextBoxInput<int>>().Single();
+                bankTextBox.Current.Value = 30;
+            });
+
+            AddAssert("volume changed", () => EditorBeatmap.HitObjects.First().Samples.All(o => o.Volume == 30));
+        }
+
+        [Test]
+        public void TestNodeSamplePointPiece()
```
Comment (peppy) [inline osu.Game.Tests/Visual/Editing/TestSceneTimelineHitObjectBlueprint.cs:190]: These tests need to clear the popover at the start or the end, currently running them more than once results in a failure.

https://github.com/ppy/osu/assets/191335/1ee0af08-bcf8-4809-bb10-c0fffbdc482f

File: osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs [hunk 4]
```diff
@@ -109,33 +151,48 @@ private void load()
...
-                volume.Current.BindValueChanged(val => updateVolumeFor(relevantObjects, val.NewValue));
+                updateAdditionBankPlaceholderText();
+                updateAdditionBankText();
+                updateAdditionBankActivated();
+                additionBank.Current.BindValueChanged(val =>
+                {
+                    updateAdditionBank(val.NewValue);
+                    updateAdditionBankPlaceholderText();
+                });
+                additionBank.OnCommit += (_, _) => updateAdditionBankText();
```
Comment (peppy) [inline osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:189]: Can logic around here be simplified like this?

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs b/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
index 37a6fa8a22..f9e04a8ca7 100644
--- a/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
@@ -178,13 +178,12 @@ private void load()
                 // this ensures that committing empty text causes a revert to the previous value.
                 bank.OnCommit += (_, _) => bank.Current.Value = getCommonBank();
 
-                updateAdditionBankPlaceholderText();
                 updateAdditionBankText();
                 updateAdditionBankActivated();
                 additionBank.Current.BindValueChanged(val =>
                 {
                     updateAdditionBank(val.NewValue);
-                    updateAdditionBankPlaceholderText();
+                    updateAdditionBankText();
                 });
                 additionBank.OnCommit += (_, _) => updateAdditionBankText();
 
@@ -257,12 +256,6 @@ private void updateBankPlaceholderText()
                 bank.PlaceholderText = string.IsNullOrEmpty(commonBank) ? "(multiple)" : string.Empty;
             }
 
-            private void updateAdditionBankPlaceholderText()
-            {
-                string? commonAdditionBank = getCommonAdditionBank();
-                additionBank.PlaceholderText = string.IsNullOrEmpty(commonAdditionBank) ? "(multiple)" : string.Empty;
-            }
-
             private void updateAdditionBankActivated()
             {
                 bool anyAdditions = relevantSamples.Any(o => o.Any(s => s.Name != HitSampleInfo.HIT_NORMAL));
@@ -280,6 +273,7 @@ private void updateAdditionBankText()
                 if (string.IsNullOrEmpty(commonAdditionBank)) return;
 
                 additionBank.Current.Value = commonAdditionBank;
+                additionBank.PlaceholderText = string.IsNullOrEmpty(commonAdditionBank) ? "(multiple)" : string.Empty;
             }
 
             private void updateVolume(int? newVolume)
```

I have more to say about this code, but this would be a first step in the right direction.

Comment (OliBomby) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:189]: Updating the text in the `ValueChanged` event has a weird side-effect that it will never allow you to clear the textbox, because changes do not propagate when the textbox is empty. I'll find a way to simplify it in a different way

File: osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs (hunk 1 above)
Comment (OliBomby) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:208]: the update method is meant to work as a little macro to reduce code duplication in all the specific update methods. Would it be clear enough if i rename some variables and add xmldoc to clarify how the flow works? I think it's quite intuitive to use once you know what it means.

**Comment by OliBomby** [comment]
Any idea why the `TestNodeSamplePointPiece` test keeps failing in the CI but locally it passes without issue?

**Comment by peppy** [comment]
It fails locally for me too. Maybe it's not for you because you're not running the other tests in the class in sequence? Sounds like something isn't being reset correctly between test methods.

JetBrains Rider 2023-06-01 at 11 32 22

**Comment by OliBomby** [comment]
I somehow failed to notice `TestSceneHitObjectSamplePointAdjustments` exists. I should move the new tests to there.

**Comment by peppy** [comment]
> I'm aware that there are always two sample point pieces overlapping at the sliderhead. I'm not really sure what to do about that. You could make them narrower and put them next to eachother like in my follow-up PR, or merge them into one piece somehow, or just remove the piece for the sliderbody bank entirely.

I'm not sure where the purple came from, but it seems weird. If you're wanting to differentiate, I think they should be different pink shades at most.

**Comment by OliBomby** [comment]
> I'm not sure where the purple came from, but it seems weird. If you're wanting to differentiate, I think they should be different pink shades at most.

I've changed it to dark pink for the sliderbody hitsound and normal pink for the node samples. Also changed the positions a bit so the node samples are always on the correct X position. I think it looks good now.

osu Game Rulesets Osu Tests_42dLfyNCnf

**Comment by OliBomby** [comment]
BTW I addressed all the feedback so the PR is ready for merge

**Comment by Tais993** [comment]
Quick bump, no reviews yet?

**Comment by bdach** [comment]
@Tais993 work not on https://github.com/orgs/ppy/projects/13 is currently being heavily deprioritised. I hope you understand.

**Comment by peppy** [comment]
@OliBomby wanna fix conflicts on this one just to keep it up-to-date?

**Comment by OliBomby** [comment]
> @OliBomby wanna fix conflicts on this one just to keep it up-to-date?

Sure

**Comment by bdach** [comment]
@OliBomby I fixed conflicts here, please double check resolution.

File: osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs (hunk 4 above)
Comment (bdach) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:195]: Not much has simplified here. Speaking from personal experience (and I've made the same mistake myself), having multiple update methods like `updateAdditionBankText` and `updateAdditionBankVisual` and `updateAdditionBank` (with a bool) is not good news. Ideally you'd want just one `updateState` that gets everything correct.

Can you explain what the issues are here *before* I try and go in with a hammer to simplify this?

File: osu.Game/Rulesets/Objects/HitObject.cs [hunk 5]
```diff
@@ -223,8 +223,16 @@ public IList<HitSampleInfo> CreateSlidingSamples()
...
         public HitSampleInfo CreateHitSampleInfo(string sampleName = HitSampleInfo.HIT_NORMAL)
         {
-            if (Samples.FirstOrDefault(s => s.Name == HitSampleInfo.HIT_NORMAL) is HitSampleInfo existingSample)
-                return existingSample.With(newName: sampleName);
+            // As per stable, all non-normal "addition" samples should use the same bank.
+            if (sampleName != HitSampleInfo.HIT_NORMAL)
+            {
+                if (Samples.FirstOrDefault(s => s.Name != HitSampleInfo.HIT_NORMAL) is HitSampleInfo existingAddition)
+                    return existingAddition.With(newName: sampleName);
+            }
```
[the remark this answers is not in the document]
Comment (OliBomby) [reply osu.Game/Rulesets/Objects/HitObject.cs:231]: This change was made by @peppy 

It seems to be a code duplication cleanup for logic that determines the correct bank to apply to a new hitsample. For example if you are making a new hitwhistle sample which is an 'addition', then this code tries to find an existing addition sample to take the bank name of.

In stable, all addition samples have the same bank, because you can only set one bank for all addition samples. This doesn't have to be true in lazer, but currently lazer behaves the same as stable for this in that you can only set one bank for all addition samples.

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Rulesets/Objects/HitObject.cs:231]: Duplication from where though? If it was deduplication I'd expect the same chunk of code to be gone from somewhere else. But I haven't seen any of that in this diff.

[the remark this answers is not in the document]
Comment (OliBomby) [reply osu.Game/Rulesets/Objects/HitObject.cs:231]: Read https://github.com/ppy/osu/pull/23443#issuecomment-1567761667 for the reasoning.
I think this PR introduced the need for the chunk of code that was duplicated and subsequently deduplicated, that's why you don't see it in the final diff.

File: osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs (hunk 4 above)
Comment (OliBomby) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:195]: I'll explain the issues as thoroughly as possible;

The update methods `updateBank`, `updateAdditionBank`, `updateVolume`, `addHitSample`, and `removeHitSample` must all only trigger when the user edits their respective value, because these are quite destructive in a sense that they synchronize the property across all selected hit objects.

Visuals must be updated before the bindings are created that update the hitobject properties on change, to prevent the hitobjects from being changed immediately when opening the popup.

Properties must be updated immediately on change, to allow the visuals in the timeline to reflect real-time changes.

Placeholder texts must be updated on change, to prevent it from incorrectly showing "(multiple)" after the user changes a property on multiple objects and clears the textbox.

The text of the textbox must not be updated on change, because otherwise the user will be unable to temporarily clear the textbox, because that is not a valid value.

File: osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs [hunk 6]
```diff
@@ -50,28 +55,54 @@ protected override bool OnClick(ClickEvent e)
...
+            return samples.FirstOrDefault(o => o.Name != HitSampleInfo.HIT_NORMAL)?.Bank ?? GetBankValue(samples);
         }
 
         public static int GetVolumeValue(ICollection<HitSampleInfo> samples)
         {
             return samples.Count == 0 ? 0 : samples.Max(o => o.Volume);
         }
 
-        public Popover GetPopover() => new SampleEditPopover(HitObject);
+        protected virtual IList<HitSampleInfo> GetSamples() => HitObject.Samples;
```
[the remark this answers is not in the document]
Comment (OliBomby) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:93]: It seems I addressed it by adding xmldoc to `SampleEditPopoer.GetRelevantSamples()`, but now I've added it to `GetSamples` too

File: osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs [hunk 7]
```diff
@@ -236,6 +244,27 @@ private void updateRepeats(IHasRepeats repeats)
...
                 });
             }
+
+            // Add node sample pieces
+            nodeSamplesContainer?.Expire();
+
+            sampleComponents.Add(nodeSamplesContainer = new Container
+            {
+                RelativeSizeAxes = Axes.Both,
+            });
```
[the remark this answers is not in the document]
Comment (OliBomby) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs]: Just clearing contents seems to work fine

File: osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs (hunk 4 above)
Comment (bdach) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs]: I think I managed to straighten this out by hoisting some null checks out of the set methods (see https://github.com/ppy/osu/pull/23443/commits/0efa028e0a82dc192303e2b99bead7ce1db66280). Please check whether this preserves all behaviours you wanted to, tests are passing as far as I could tell.

File: osu.Game/Rulesets/Objects/HitObject.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Rulesets/Objects/HitObject.cs:231]: Well the thing is that I'm not confident that it's a safe change to make. Yes `CreateHitSampleInfo()` is frequently used in editor contexts, but it also is frequently used in *non*-editor context. Such as:

https://github.com/ppy/osu/blob/a8a2e543d51788b1dc05ab318b54b46af76d0be7/osu.Game.Rulesets.Osu/Objects/Spinner.cs#L94
https://github.com/ppy/osu/blob/a8a2e543d51788b1dc05ab318b54b46af76d0be7/osu.Game.Rulesets.Taiko/UI/DrumSampleTriggerSource.cs#L43-L52

So as far as I can tell this probably changes sample playback behaviour for some objects. So I'd like to know whether this was a change made specifically for the editor flows, or something that is supposed to be valid for everything everywhere.

@peppy mind providing context / commentary on this when able?

**Comment by bdach** [comment]
> I've changed it to dark pink for the sliderbody hitsound and normal pink for the node samples.

I felt that the contrast with this between the dark pink bg and the black text was pretty dreadful so I lightened it up a touch in https://github.com/ppy/osu/pull/23443/commits/7d5dc750e53938d60fa6cffedc1879d32c546118:



The only remaining concern to me with this is [this review thread](https://github.com/ppy/osu/pull/23443#discussion_r1629371984). If we can revert this change / localise it to the editor, or justify its correctness in all contexts _somehow_, then I think this is good to go.

**Comment by OliBomby** [comment]
I like the change to the contrast of the alternative colour.

> The only remaining concern to me with this is https://github.com/ppy/osu/pull/23443#discussion_r1629371984. If we can revert this change / localise it to the editor, or justify its correctness in all contexts somehow, then I think this is good to go.

I'll look into it to see what this change exactly means for all the usages of this method.

File: osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs (hunk 4 above)
Comment (OliBomby) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs]: Looks good. I did some small cleanup and fixed an unrelated bug i found while testing the behaviours.

File: osu.Game/Rulesets/Objects/HitObject.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (OliBomby) [reply osu.Game/Rulesets/Objects/HitObject.cs:231]: I did some analysis:

It seems the change allows SpinnerBonusTick to receive an addition bank if the spinner has additions.
This is probably incorrect because spinnerbonus-max and spinnerspin seem to both always use the primary bank.

For taiko Hit and StrongableHitObject, this change allows clap and finish samples added by the Type and IsStrong bindables to use the addition bank.
I think this change is desirable in this case because the user is able to change the addition bank (since this PR), so probably expects all the whistle, clap, and finish samples to have the same addition bank. It would be weird if for example the user sets the clap sample to the normal bank, sets IsStrong to true, and finds a finish sample with the soft bank from the primary bank.

For DrumSampleTriggerSource, this change makes it possible to play the rim and centre samples on different banks if the most valid object has different primary and addition banks. It adds a little bit of mapper intention but maybe we don't want this.

I'm not familiar with taiko hitsounding, but it seems like taiko always plays a single sample per hit so having two different banks on the same object might be a silly idea. If that's the case then we should probably go back and always take the bank of the hitnormal in taiko.

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Rulesets/Objects/HitObject.cs:231]: So I looked into this myself a bit further and well.....

Here's a test map that I did: [Moderat - Rusty Nails - Copy.zip](https://github.com/user-attachments/files/15817315/Moderat.-.Rusty.Nails.-.Copy.zip)

The thing to note here is that I had all of the sample files just be a speech synthesizer that says the name of the bank. As such:

- For the spinner thing, now I believe it to be an irrelevant concern because stable did not support banks for `spinnerbonus` at all. So the fact this even works is a lazer-only thing and not really advertised as supported anywhere, so I'm fine with leaving it as "undefined behaviour".
- For the taiko thing... well let's just say neither `master` nor this PR follow what stable does.
  - On stable, I can't get the drum to produce anything audible other than "normal". Which means stable does not appear to care about sample bank on individual objects *at all*. The only thing that appears to have an effect there is the thing "default sample settings" (it'll be all normal, all soft, or all drum, always).
  - On `master`, every object with additions will follow its "primary" / `hitnormal` bank,
  - while on this branch, every object with additions will follow its "addition" bank.

I'm not sure what to think of the taiko thing. I'll go ask the taiko NAT maybe for some advice because I'm stupid and know nothing about mapping, but I'm not sure this needs to be a blocking concern for this PR.

With that said I'm going to just probably be looking to get this in and address any issues later. @ppy/team-client any objections to that?

**Review by bdach — APPROVED** [review]

File: osu.Game.Rulesets.Catch/Objects/JuiceStream.cs [hunk 8]
```diff
@@ -76,6 +76,8 @@ protected override void CreateNestedHitObjects(CancellationToken cancellationTok
         {
             base.CreateNestedHitObjects(cancellationToken);
 
+            this.PopulateNodeSamples();
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Catch/Objects/JuiceStream.cs:79]: So is every single implementation of `CreateNestedHitObjects` going to have to call this?

Comment (bdach) [reply osu.Game.Rulesets.Catch/Objects/JuiceStream.cs:79]: Every `IHasRepeats` would need to I suppose. Not entirely sure how to fix that without introducing eldritch abstract entities.

**Comment by bdach** [comment]
One other thing I didn't really consider (which was pointed out to me in a [reddit discussion](https://www.reddit.com/r/osugame/comments/1dga6m5/comment/l8s0gms/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button)) is that this UI does not work with buzz sliders:




@OliBomby thoughts on how we can remedy this?

**Comment by OliBomby** [comment]
I can think of a few workarounds but the UX leaves a lot to be desired. As a workaround you can zoom in really far in the timeline to create more space so the sample points dont overlap anymore.
You could also multiselect another slider and edit the body hitsounds of the other to hit the buzz slider body at the same time.

The information density of buzz sliders is pretty hard to handle on the timeline. Maybe a fix could be to merge the sample point pieces into a single expanding box whenever they overlap too much. Clicking it would then expend it down vertically and show every node to be edited. This still would work for super long buzz sliders though.

**Comment by bdach** [comment]
The more these sorts of issues pop up, the more I think we need a new dedicated screen for hitsounding. A few of the ideas you proposed popped up in my head too but I'm not super sure they are going to work well in practice.

File: osu.Game.Rulesets.Catch/Objects/JuiceStream.cs (hunk 8 above)
Comment (peppy) [reply osu.Game.Rulesets.Catch/Objects/JuiceStream.cs:79]: @smoogipoo do you have strong opinions on this?

**Comment by OliBomby** [comment]
> The more these sorts of issues pop up, the more I think we need a new dedicated screen for hitsounding. A few of the ideas you proposed popped up in my head too but I'm not super sure they are going to work well in practice.

I want to add a dedicated screen for hitsounding too, alongside being able to do basic hitsounds in the compose screen. This is because [some people prefer hitsounding while mapping](https://x.com/OliBomby/status/1637856574848000001). 

A dedicated screen for hitsounding will definitely make hitsounding a lot better of an experience for anything that isn't just adding simple whistle, finish, claps.

File: osu.Game.Rulesets.Catch/Objects/JuiceStream.cs (hunk 8 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Catch/Objects/JuiceStream.cs:79]: not at the moment, no

File: osu.Game/Rulesets/Objects/HitObject.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Rulesets/Objects/HitObject.cs:231]: > I'm not sure what to think of the taiko thing.

As far as I've experienced, taiko players they expect to hear a taiko drum sound and hitsounds either take a backseat being "whatever" or (as we did in lazer) are more present to the point of annoying them.

> With that said I'm going to just probably be looking to get this in and address any issues later. @ppy/team-client any objections to that?

Sounds fine.

**Review by peppy — APPROVED** [review]
